org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java

```diff
@@ -0,0 +1,214 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2025 Zsombor Gegesy and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Zsombor Gegesy - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.internal.debug.ui;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.debug.core.DebugException;
+import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.debug.core.IJavaStackFrame;
+import org.eclipse.jdt.debug.core.IJavaStackFrame.Category;
+import org.eclipse.jdt.internal.ui.JavaPluginImages;
+import org.eclipse.jdt.internal.ui.filtertable.FilterManager;
+import org.eclipse.jface.preference.IPreferenceStore;
+import org.eclipse.jface.resource.ImageDescriptor;
+import org.eclipse.jface.util.IPropertyChangeListener;
+import org.eclipse.jface.util.PropertyChangeEvent;
+
+/**
+ * Provides foreground and background colors for stack frames in a debug view. After usage, it needs to be closed, so it could unregister itself from
+ * preference storage.
+ *
+ */
+public final class StackFramePresentationProvider implements IPropertyChangeListener {
+
+	public static final FilterManager PLATFORM_STACK_FRAMES = new FilterManager(IJDIPreferencesConstants.PREF_ACTIVE_PLATFORM_FRAME_FILTER_LIST, IJDIPreferencesConstants.PREF_INACTIVE_PLATFORM_FRAME_FILTER_LIST);
+	public static final FilterManager CUSTOM_STACK_FRAMES = new FilterManager(IJDIPreferencesConstants.PREF_ACTIVE_CUSTOM_FRAME_FILTER_LIST, IJDIPreferencesConstants.PREF_INACTIVE_CUSTOM_FRAME_FILTER_LIST);
+
+	/**
+	 * Class to decide if a particular class name is part of a list of classes and list of packages.
+	 */
+	static class Filters {
+		private final String[] filters;
+
+		Filters(String[] filters) {
+			this.filters = filters;
+		}
+
+		boolean match(String fqcName) {
+			for (String filter : filters) {
+				if (filter.endsWith("*")) { //$NON-NLS-1$
+					if (fqcName.startsWith(filter.substring(0, filter.length() - 1))) {
+						return true;
+					}
+				} else {
+					if (filter.equals(fqcName)) {
+						return true;
+					}
+				}
+			}
+			return false;
+		}
+	}
+
+	private Filters platform;
```

Comment:
can those be declared final or even use a record?

Reply:
The variables inside StackFramePresentationProvider are needs to be mutable, but Filters can be turned into a record.